File: app/src/main/java/com/tunjid/fingergestures/gestureconsumers/RotationGestureConsumer.java
package com.tunjid.fingergestures.gestureconsumers;


import android.content.Intent;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.provider.Settings;
import android.view.accessibility.AccessibilityEvent;

import com.tunjid.fingergestures.App;
import com.tunjid.fingergestures.R;
import com.tunjid.fingergestures.SetManager;
import com.tunjid.fingergestures.billing.PurchasesManager;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.util.Comparator;
import java.util.List;
import java.util.Set;

import androidx.annotation.Nullable;
import androidx.annotation.StringDef;

import static android.view.accessibility.AccessibilityEvent.TYPE_WINDOW_STATE_CHANGED;
import static com.tunjid.fingergestures.App.withApp;
import static com.tunjid.fingergestures.services.FingerGestureService.ANDROID_SYSTEM_UI_PACKAGE;

public class RotationGestureConsumer implements GestureConsumer {

    private static final int ENABLE_AUTO_ROTATION = 1;
    private static final int DISABLE_AUTO_ROTATION = 0;

    public static final String ACTION_WATCH_WINDOW_CHANGES = "com.tunjid.fingergestures.watch_windows";
    public static final String EXTRA_WATCHES_WINDOWS = "extra watches window content";
    public static final String EXCLUDED_APPS = "excluded rotation apps";
    public static final String ROTATION_APPS = "rotation apps";
    private static final String WATCHES_WINDOW_CONTENT = "watches window content";
    private static final String EMPTY_STRING = "";

    @Retention(RetentionPolicy.SOURCE)
    @StringDef({ROTATION_APPS, EXCLUDED_APPS})
    public @interface PersistedSet {}

    private final SetManager<ApplicationInfo> setManager;

    private String lastPackageName;

    private static RotationGestureConsumer instance;

    private RotationGestureConsumer() {
        setManager = new SetManager<>(this::compareApplicationInfo, this::canAddToSet, this::fromPackageName, this::fromApplicationInfo);
        setManager.addToSet(ANDROID_SYSTEM_UI_PACKAGE, EXCLUDED_APPS);
        withApp(app -> setManager.addToSet(app.getPackageName(), EXCLUDED_APPS));
    }

    public static RotationGestureConsumer getInstance() {
        if (instance == null) instance = new RotationGestureConsumer();
        return instance;
    }

    @Override
    public void onGestureActionTriggered(int gestureAction) {
        setAutoRotateOn(!isAutoRotateOn());
    }

    @Override
    public boolean accepts(int gesture) {
        return gesture == TOGGLE_AUTO_ROTATE;
    }

    public void onAccessibilityEvent(AccessibilityEvent event) {
        if (event == null) return;
        if (!App.canWriteToSettings() || event.getEventType() != TYPE_WINDOW_STATE_CHANGED) return;

        CharSequence sequence = event.getPackageName();
        if (sequence == null) return;

        String packageName = sequence.toString();
        if (packageName.equals(lastPackageName)) return;

        Set<String> rotationApps = setManager.getSet(ROTATION_APPS);
        if (rotationApps.isEmpty() || setManager.getSet(EXCLUDED_APPS).contains(packageName))
            return;

        lastPackageName = packageName;
        setAutoRotateOn(rotationApps.contains(packageName));
    }

    public String getAddText(@PersistedSet String preferencesName) {
        return App.transformApp(app -> app.getString(R.string.auto_rotate_add, ROTATION_APPS.equals(preferencesName)
                ? app.getString(R.string.auto_rotate_apps)
                : app.getString(R.string.auto_rotate_apps_excluded)), EMPTY_STRING);
    }

    public String getRemoveText(@PersistedSet String preferencesName) {
        return App.transformApp(app -> app.getString(R.string.auto_rotate_remove, ROTATION_APPS.equals(preferencesName)
                ? app.getString(R.string.auto_rotate_apps)
                : app.getString(R.string.auto_rotate_apps_excluded)), EMPTY_STRING);
    }

    public boolean isRemovable(String packageName) {
        return App.transformApp(app -> (!ANDROID_SYSTEM_UI_PACKAGE.equals(packageName) && !app.getPackageName().equals(packageName)), false);
    }

    public boolean addToSet(String packageName, @PersistedSet String preferencesName) {
        return setManager.addToSet(packageName, preferencesName);
    }

    public void removeFromSet(String packageName, @PersistedSet String preferencesName) {
        setManager.removeFromSet(packageName, preferencesName);
    }

    public List<ApplicationInfo> getList(@PersistedSet String preferenceName) {
        return setManager.getItems(preferenceName);
    }

    public Comparator<ApplicationInfo> getApplicationInfoComparator() {
        return this::compareApplicationInfo;
    }

    public boolean canAutoRotate() {
        return App.transformApp(app -> app.getPreferences().getBoolean(WATCHES_WINDOW_CONTENT, false), false);
    }

    public void enableWindowContentWatching(boolean enabled) {
        withApp(app -> {
            app.getPreferences().edit().putBoolean(WATCHES_WINDOW_CONTENT, enabled).apply();

            Intent intent = new Intent(ACTION_WATCH_WINDOW_CHANGES);
            intent.putExtra(EXTRA_WATCHES_WINDOWS, enabled);

            app.broadcast(intent);
        });
    }

    private void setAutoRotateOn(boolean isOn) {
        withApp(app -> {
            int enabled = isOn ? ENABLE_AUTO_ROTATION : DISABLE_AUTO_ROTATION;
            Settings.System.putInt(app.getContentResolver(), Settings.System.ACCELEROMETER_ROTATION, enabled);
        });
    }

    private boolean isAutoRotateOn() {
        return App.transformApp(app -> Settings.System.getInt(app.getContentResolver(), Settings.System.ACCELEROMETER_ROTATION, DISABLE_AUTO_ROTATION) == ENABLE_AUTO_ROTATION, false);
    }

    private boolean canAddToSet(String preferenceName) {
        Set<String> set = setManager.getSet(preferenceName);
        long count = set.stream().filter(this::isRemovable).count();
        return count < 2 || PurchasesManager.getInstance().isPremiumNotTrial();
    }

    private int compareApplicationInfo(ApplicationInfo infoA, ApplicationInfo infoB) {
        return App.transformApp(app -> {
            PackageManager packageManager = app.getPackageManager();
            return packageManager.getApplicationLabel(infoA).toString().compareTo(packageManager.getApplicationLabel(infoB).toString());
        }, 0);
    }

    private String fromApplicationInfo(ApplicationInfo info) {
        return info.packageName;
    }

    @Nullable
    private ApplicationInfo fromPackageName(String packageName) {
        return App.transformApp(app -> {
            try {return app.getPackageManager().getApplicationInfo(packageName, 0);}
            catch (Exception e) {e.printStackTrace();}

            return null;
        }, null);
    }
}

File: app/src/main/java/com/tunjid/fingergestures/gestureconsumers/RotationGestureConsumer.kt
/*
 * Copyright (c) 2017, 2018, 2019 Adetunji Dahunsi.
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.tunjid.fingergestures.gestureconsumers


import android.content.pm.ApplicationInfo
import android.os.Parcelable
import android.provider.Settings
import android.view.accessibility.AccessibilityEvent
import android.view.accessibility.AccessibilityEvent.TYPE_WINDOW_STATE_CHANGED
import com.tunjid.fingergestures.*
import com.tunjid.fingergestures.billing.PurchasesManager
import com.tunjid.fingergestures.di.AppBroadcaster
import com.tunjid.fingergestures.models.Broadcast
import com.tunjid.fingergestures.services.FingerGestureService.Companion.ANDROID_SYSTEM_UI_PACKAGE
import io.reactivex.Flowable
import io.reactivex.processors.BehaviorProcessor
import io.reactivex.schedulers.Schedulers
import kotlinx.parcelize.Parcelize
import java.util.*
import javax.inject.Inject

class RotationGestureConsumer @Inject constructor(
    private val app: App,
    private val broadcaster: AppBroadcaster,
    private val purchasesManager: PurchasesManager
) : GestureConsumer {

    @Parcelize
    enum class Preference(override val preferenceName: String) : SetPreference, Parcelable {
        RotatingApps(preferenceName = "rotation apps"),
        NonRotatingApps(preferenceName = "excluded rotation apps"),
    }

    val setManager: SetManager<Preference, ApplicationInfo> = SetManager(
        keys = Preference.values().asIterable(),
        sorter = Comparator(this::compareApplicationInfo),
        addFilter = this::canAddToSet,
        stringMapper = this::fromPackageName,
        objectMapper = ApplicationInfo::packageName
    )

    val autoRotatePreference: ReactivePreference<Boolean> = ReactivePreference(
        preferencesName = WATCHES_WINDOW_CONTENT,
        default = false,
        onSet = { broadcaster(Broadcast.Service.WatchesWindows(enabled = it)) }
    )

    val applicationInfoComparator: Comparator<ApplicationInfo>
        get() = Comparator { infoA, infoB -> this.compareApplicationInfo(infoA, infoB) }

    val lastSeenApps
        get() = shifter.flowable.map { it.mapNotNull(this::fromPackageName) }
            .subscribeOn(Schedulers.io())

    val unRemovablePackages = listOf(
        ANDROID_SYSTEM_UI_PACKAGE,
        app.packageName
    )

    private var lastPackageName: String? = null
    private val shifter = Shifter(9)

    private var isAutoRotateOn: Boolean
        get() = App.transformApp({ app ->
            Settings.System.getInt(
                app.contentResolver,
                Settings.System.ACCELEROMETER_ROTATION,
                DISABLE_AUTO_ROTATION) == ENABLE_AUTO_ROTATION
        }, false)
        set(isOn) {
            App.withApp { app ->
                val enabled = if (isOn) ENABLE_AUTO_ROTATION else DISABLE_AUTO_ROTATION
                Settings.System.putInt(app.contentResolver, Settings.System.ACCELEROMETER_ROTATION, enabled)
            }
        }

    init {
        App.withApp { app ->
            setManager.editorFor(Preference.NonRotatingApps).apply {
                plus(ApplicationInfo().apply { packageName = ANDROID_SYSTEM_UI_PACKAGE })
                plus(app.applicationInfo)
            }
        }
    }

    override fun onGestureActionTriggered(gestureAction: GestureAction) {
        isAutoRotateOn = !isAutoRotateOn
    }

    override fun accepts(gesture: GestureAction): Boolean = gesture == GestureAction.TOGGLE_AUTO_ROTATE

    fun onAccessibilityEvent(event: AccessibilityEvent?) {
        if (event == null) return
        if (!App.canWriteToSettings || event.eventType != TYPE_WINDOW_STATE_CHANGED) return

        val sequence = event.packageName ?: return

        val packageName = sequence.toString()
        if (packageName == lastPackageName) return

        val rotationApps = setManager.getSet(Preference.RotatingApps)
        if (rotationApps.isEmpty() || setManager.getSet(Preference.NonRotatingApps).contains(packageName))
            return

        lastPackageName = packageName
        shifter + packageName
        isAutoRotateOn = rotationApps.contains(packageName)
    }

    fun getAddText(preferencesName: Preference): String = App.transformApp({ app ->
        app.getString(R.string.auto_rotate_add, if (Preference.RotatingApps == preferencesName)
            app.getString(R.string.auto_rotate_apps)
        else
            app.getString(R.string.auto_rotate_apps_excluded))
    }, EMPTY_STRING)

    fun getRemoveText(preferencesName: Preference): String = App.transformApp({ app ->
        app.getString(R.string.auto_rotate_remove, if (Preference.RotatingApps == preferencesName)
            app.getString(R.string.auto_rotate_apps)
        else
            app.getString(R.string.auto_rotate_apps_excluded))
    }, EMPTY_STRING)

    private fun canAddToSet(preferenceName: Preference): Boolean {
        val set = setManager.getSet(preferenceName)
        val count = set.filterNot(unRemovablePackages::contains).count()
        return count < 2 || purchasesManager.isPremiumNotTrial
    }

    private fun compareApplicationInfo(infoA: ApplicationInfo, infoB: ApplicationInfo): Int {
        val packageManager = app.packageManager
        return packageManager.getApplicationLabel(infoA).toString()
            .compareTo(packageManager.getApplicationLabel(infoB).toString())
    }

    private fun fromPackageName(packageName: String): ApplicationInfo? = App.transformApp { app ->
        try {
            app.packageManager.getApplicationInfo(packageName, 0)
        } catch (e: Exception) {
            e.printStackTrace()
            null
        }
    }

    companion object {
        private const val ENABLE_AUTO_ROTATION = 1
        private const val DISABLE_AUTO_ROTATION = 0
        private const val WATCHES_WINDOW_CONTENT = "watches window content"
        private const val EMPTY_STRING = ""
    }
}

private class Shifter(private val size: Int) {
    private val backing = arrayOfNulls<String>(size)
    private val watcher = BehaviorProcessor.create<List<String>>()

    val flowable: Flowable<List<String>> = watcher.doOnSubscribe { push() }

    operator fun plus(value: String) {
        System.arraycopy(backing, 0, backing, 1, size - 1)
        backing[0] = value
        if (watcher.hasSubscribers()) push()
    }

    private fun push() = watcher.onNext(backing.filterNotNull())
}